src/main/java/org/apache/sysml/runtime/instructions/cp/BuiltinUnaryCPInstruction.java

```diff
@@ -21,57 +21,92 @@
 import org.apache.sysml.parser.Expression.ValueType;
 import org.apache.sysml.runtime.DMLRuntimeException;
+import org.apache.sysml.runtime.controlprogram.caching.MatrixObject;
 import org.apache.sysml.runtime.functionobjects.Builtin;
 import org.apache.sysml.runtime.functionobjects.ValueFunction;
 import org.apache.sysml.runtime.instructions.InstructionUtils;
+import org.apache.sysml.runtime.matrix.data.LibCommonsMath;
+import org.apache.sysml.runtime.matrix.data.MatrixBlock;
 import org.apache.sysml.runtime.matrix.operators.Operator;
 import org.apache.sysml.runtime.matrix.operators.SimpleOperator;
 import org.apache.sysml.runtime.matrix.operators.UnaryOperator;
+import org.apache.sysml.runtime.controlprogram.context.ExecutionContext;
 
-public abstract class BuiltinUnaryCPInstruction extends UnaryCPInstruction {
+import static org.apache.sysml.runtime.instructions.InstructionUtils.isBuiltinFunction;
+
+public class BuiltinUnaryCPInstruction extends UnaryCPInstruction {
+
+	protected CPOperand _out;
 
 	protected BuiltinUnaryCPInstruction(Operator op, CPOperand in, CPOperand out, String opcode,
-			String istr) {
+										String istr) {
 		super(CPType.BuiltinUnary, op, in, out, opcode, istr);
+		_out = out;
 	}
 
-	public static BuiltinUnaryCPInstruction parseInstruction ( String str ) 
-		throws DMLRuntimeException 
+	public static BuiltinUnaryCPInstruction parseInstruction ( String str )
+			throws DMLRuntimeException
 	{
 		CPOperand in = new CPOperand("", ValueType.UNKNOWN, DataType.UNKNOWN);
 		CPOperand out = new CPOperand("", ValueType.UNKNOWN, DataType.UNKNOWN);
-		
+
 		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
 		String opcode = null;
+
 		ValueFunction func = null;
-		
+
 		//print or stop or cumulative aggregates
-		if( parts.length==4 ) 
+		if( parts.length==4 )
 		{
 			opcode = parts[0];
 			in.split(parts[1]);
 			out.split(parts[2]);
 			func = Builtin.getBuiltinFnObject(opcode);
-			
+
 			if( Arrays.asList(new String[]{"ucumk+","ucum*","ucummin","ucummax"}).contains(opcode) )
-				return new MatrixBuiltinCPInstruction(new UnaryOperator(func,Integer.parseInt(parts[3])), in, out, opcode, str); 
+				return new MatrixBuiltinCPInstruction(new UnaryOperator(func,Integer.parseInt(parts[3])), in, out, opcode, str);
 			else
 				return new ScalarBuiltinCPInstruction(new SimpleOperator(func), in, out, opcode, str);
 		}
 		else //2+1, general case
 		{
 			opcode = parseUnaryInstruction(str, in, out);
-			func = Builtin.getBuiltinFnObject(opcode);
-			
-			if(in.getDataType() == DataType.SCALAR)
-				return new ScalarBuiltinCPInstruction(new SimpleOperator(func), in, out, opcode, str);
-			else if(in.getDataType() == DataType.MATRIX)
-				return new MatrixBuiltinCPInstruction(new UnaryOperator(func), in, out, opcode, str);
+			if(LibCommonsMath.isSupportedUnaryOperation(opcode)) {
+				return new BuiltinUnaryCPInstruction(null, in, out, opcode, str);
+			}
```

Comment:
Why is this modification required? The existing `MatrixBuiltinCPInstruction `should directly apply here as well, right?

Reply:
Sorry, please consider modifying if it could work. Infact, cholesky worked well a month ago. Now, there seems to be some problem with this so I came this way. Thanks.